ops/hookcmds.py

```diff
@@ -0,0 +1,995 @@
+
+
+def state_delete(key: str):
+    """Delete server-side-state key value pairs."""
+    _run('state-delete', key)
+
+
+@overload
+def state_get(key: str) -> str: ...
+@overload
+def state_get(key: None) -> dict[str, str]: ...
+def state_get(key: str | None) -> dict[str, str] | str:
+    """Get server-side-state value."""
+    args = ['state-get']
+    if key is not None:
+        args.append(key)
+    result = _run(*args, return_output=True, use_json=True)
+    return cast('dict[str, str]', result) if key is None else cast('str', result)
+
+
+def state_set(data: Mapping[str, str], file: pathlib.Path | None = None):
+    """Set server-side-state values."""
+    args = ['state-set']
+    args.extend(f'{k}={v}' for k, v in data.items())
+    if file is not None:
+        args.extend(['--file', str(file)])
+    _run(*args)
+
+
+@overload
+def status_get(*, include_data: Literal[True], app: Literal[False]) -> UnitStatus: ...
+@overload
+def status_get(*, include_data: Literal[True], app: Literal[True]) -> AppStatus: ...
+@overload
+def status_get(*, include_data: Literal[False], app: bool = False) -> str: ...
+def status_get(*, include_data: bool = False, app: bool = False) -> AppStatus | UnitStatus | str:
+    """Get a status of a unit or an application."""
+    args = ['status-get', f'--application={str(app).lower()}']
+    if include_data:
+        args.append('--include-data')
+    result = _run(*args, use_json=True, return_output=True)
+    if app:
+        app_status = cast('_AppStatusDict', result['application-status'])
+        units = {
+            name: UnitStatus(
+                status=u['status'], message=u['message'], status_data=u['status-data']
+            )
+            for name, u in app_status.get('units', {}).items()
+        }
+        return AppStatus(
+            status=app_status['status'],
+            message=app_status['message'],
+            status_data=app_status['status-data'],
+            units=units,
+        )
+    unit_status = cast('_UnitStatusDict', result)
+    return UnitStatus(
+        status=unit_status['status'],
+        message=unit_status['message'],
+        status_data=unit_status['status-data'],
+    )
+
+
+def status_set(status: SettableStatusName, message: str = '', *, app: bool = False):
+    """Set a status of a unit or an application.
+
+    Args:
+        status: The status to set.
+        message: The message to set in the status.
+        app: A boolean indicating whether the status should be set for a unit or an
+            application.
+    """
+    args = ['status-set', f'--application={app}', status]
+    if message is not None:
+        args.append(message)
+    _run(*args)
+
+
+def storage_add(name: str, count: int = 1):
+    """Add storage instances."""
+    _run('storage-add', f'{name}={count}')
+
+
+@overload
+def storage_get(identifier: str | None = None, *, attribute: str) -> str: ...
+@overload
+def storage_get(identifier: str | None = None) -> Storage: ...
```

Comment:
What does *attribute* actually do? It's a bit odd, normally on these functions providing a key/attribute would return a single str and not providing it would return a dict, so just wondering why the different here (might be a Juju quirk).